### PR TITLE
feat: cast --from-bin

### PR DESCRIPTION
**Pull request by odyslam**

### CODE REVIEW

File: cli/src/cast.rs [hunk 1]
```diff
@@ -80,6 +80,11 @@ async fn main() -> eyre::Result<()> {
             };
             println!("0x{}", output);
         }
+        Subcommands::FromBin { value } => {
+            let val = unwrap_or_stdin(value)?;
+            let n: u32 = u32::from_str_radix(&val, 2).unwrap();
```
Comment (onbjerg) [inline cli/src/cast.rs:85]: This will only allow a single 32 bit integer to be converted, whereas `seth` converts a binary blob into hex

**Review by onbjerg — CHANGES_REQUESTED** [review]

File: cli/src/cast.rs (hunk 1 above)
Comment (odyslam) [reply cli/src/cast.rs:85]: @onbjerg help out a bit here. What should be the input? I imagined a String that represents binary data, aka "101010101010101..".

Comment (onbjerg) [reply cli/src/cast.rs:85]: Yes, but the string can be larger than 32 bits. Example input could be the bytecode of a contract in binary, which could be up to 22 kilobytes of data

Comment (odyslam) [reply cli/src/cast.rs:85]: Right, awesome. Thanks for the comments, will fix it asap

**Comment by gakonst** [comment]
@odyslam any plans to follow up here?

**Comment by odyslam** [comment]
Thanks for the ping @gakonst

from-bin coming through

File: cli/src/cast.rs [hunk 2]
```diff
@@ -80,6 +80,18 @@ async fn main() -> eyre::Result<()> {
...
+            let val = unwrap_or_stdin(value)?;
+            let hex_string = val
+                .as_bytes()
+                .chunks(4)
+                .map(str::from_utf8)
+                .map(|chunk| u32::from_str_radix(&chunk.unwrap(), 2))
+                .map(|number| format!("{:x}", number.unwrap()))
+                .collect::<Vec<String>>()
+                .join("");
+            println!("0x{}", hex_string)
```
Comment (onbjerg) [inline cli/src/cast.rs:93]: This is simpler:

```suggestion
            let hex_string: String = (0..val.len())
                .step_by(8)
                .map(|i| u8::from_str_radix(&val[i..i + 8], 2).expect("Invalid binary"))
                .map(|n| format!("{:x}", n))
                .collect();
            println!("0x{}", hex_string)
```

Unfortunately no way to use `std::fmt::LowerHex` on a `Vec<u8>`, that would have been the cleanest ultimately

Comment (onbjerg) [reply cli/src/cast.rs:93]: Although I guess we should actually step by 4 (or 2?) just to be sure

Comment (odyslam) [reply cli/src/cast.rs:93]: @onbjerg if the hex is encoded with 4 bits, that doesn't mean that we **have** to step by 4?

Comment (onbjerg) [reply cli/src/cast.rs:93]: No, if we pass 8 bits you just get 2 characters. The amount we can step by is just determined by what kind of input we expect

Comment (odyslam) [reply cli/src/cast.rs]: Right, the algo will always convert 4 bits to a char. Thanks!

**Review by onbjerg — APPROVED** [review]
lgtm

**Review by gakonst — COMMENTED** [review]
I'm not sure this works?
```
echo -ne "\x12\x34" | seth --from-bin
```

works

vs
```
echo -ne "\x12\x34" | ./target/debug/cast --from-bin
The application panicked (crashed).
Message:  byte index 8 is out of bounds of `4`
Location: cli/src/cast.rs:89
```

pls review / test it out

**Comment by onbjerg** [comment]
My bad, I thought the feature was translating a string of 0s and 1s to hex, not actual binary data

**Comment by odyslam** [comment]
yy, implementing asap. Sorry for this

**Comment by onbjerg** [comment]
Hey, just wanted to check up on the status of this :smile:

**Comment by odyslam** [comment]
@onbjerg so I have been messing with it, but the outputs are not exactly the same. 

For some reason, seth adds an extra `00` in front: (e.g `0x0007` and `0x07`), while also pads the end of the file with extra zeroes. 

While their hex representation is not the same, due to the padding, I am not sure if they are information-wise the same, even if the hex string output isn't.

**Pull request closed without merging**